pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java

```diff
@@ -0,0 +1,163 @@
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.segment.spi.index.reader.Dictionary;
+import org.apache.pinot.spi.data.FieldSpec;
+
+/**
+ * Operator which executes DISTINCT operation based on dictionary
+ */
+public class DictionaryBasedDistinctOperator extends DistinctOperator {
+    private static final String OPERATOR_NAME = "DictionaryBasedDistinctOperator";
+    int MAX_INITIAL_CAPACITY = 10000;
+
+    private final DistinctAggregationFunction _distinctAggregationFunction;
+    private final Map<String, Dictionary> _dictionaryMap;
+    private final int _numTotalDocs;
+    private final TransformOperator _transformOperator;
+    private final IntOpenHashSet _dictIdSet;
+
+    private IntPriorityQueue _priorityQueue;
+
+    private boolean _hasOrderBy;
+
+    public DictionaryBasedDistinctOperator(IndexSegment indexSegment, DistinctAggregationFunction distinctAggregationFunction,
+                                           Map<String, Dictionary> dictionaryMap, int numTotalDocs,
+                                           TransformOperator transformOperator) {
+        super(indexSegment, distinctAggregationFunction, transformOperator);
+
+        _distinctAggregationFunction = distinctAggregationFunction;
+        _dictionaryMap = dictionaryMap;
+        _numTotalDocs = numTotalDocs;
+        _transformOperator = transformOperator;
+        _dictIdSet = new IntOpenHashSet();
+
+        List<OrderByExpressionContext> orderByExpressionContexts = _distinctAggregationFunction.getOrderByExpressions();
+
+        if (orderByExpressionContexts != null) {
+            OrderByExpressionContext orderByExpressionContext = orderByExpressionContexts.get(0);
+            int limit = _distinctAggregationFunction.getLimit();
+            int comparisonFactor = orderByExpressionContext.isAsc() ? -1 : 1;
+            _priorityQueue =
+                    new IntHeapPriorityQueue(Math.min(limit, MAX_INITIAL_CAPACITY), (i1, i2) -> (i1 - i2) * comparisonFactor);
+            _hasOrderBy = true;
+        }
+    }
+
+    @Override
+    protected IntermediateResultsBlock getNextBlock() {
+        int limit = _distinctAggregationFunction.getLimit();
+        String column = _distinctAggregationFunction.getInputExpressions().get(0).getIdentifier();
+
+        assert _distinctAggregationFunction.getType() == AggregationFunctionType.DISTINCT;
+
+        Dictionary dictionary = _dictionaryMap.get(column);
+        int dictionarySize = dictionary.length();
+
+        for (int dictId = 0; dictId < dictionarySize; dictId++) {
+            if (!_hasOrderBy) {
+                _dictIdSet.add(dictId);
+
+                if (_dictIdSet.size() >= limit) {
+                    break;
+                }
+            } else {
+                // We already ascertained that the dictionary is sorted, hence we can use IDs instead of actual values
```

Comment:
Since we already know the dictionary is sorted, simply pick the first (asc) or last (desc) `limit` dictIds and put them into the `DistinctTable`, no need to scan the dictionary.

Reply:
That would not work,  since dictionary does not allow us to get dictIds from it directly. We would have to get it from TransformBlock, which does not guarantee order of returned dictIds.

Reply:
I don't follow. dictId is simply the index within dictionary. E.g. for ascending order of limit 10, just read dictId 0-9 from the dictionary and construct the DistinctTable. No need to scan the whole dictionary

Reply:
Yeah, as discussed, I was not aware that dictIDs are strictly in the order of 0 - dictionary limit. Changed now, thanks